draft-ppm-protocol.md

```diff
@@ -647,147 +676,222 @@ be cryptographically protected as described in {{helper-state}}.
+  opaque aggregation_param<0..2^16-1>;
+  VerifyStartSubReq seq<1..2^24-1>;
+} VerifyStartReq;
 ~~~
 
-The structure contains the PPM task, an opaque *helper state* string, and a
-sequence of *sub-requests*, each corresponding to a unique client report.
-Sub-requests are structured as follows:
+The structure contains the PPM task, an opaque *helper state* string
+(`helper_state`), and a sequence of *sub-requests* (`seq`), each corresponding
+to a unique client report. The request also carries the aggregation parameter
+(`aggregation_param`) used for VDAF evaluation. Sub-requests are structured as
+follows:
 
 ~~~
 struct {
-  Time time;                       // Equal to Report.time.
-  uint64 nonce;                    // Equal to Report.nonce.
-  Extension extensions<4..2^16-1>; // Equal to Report.extensions.
+  Time time;                       // Report.time
+  uint64 nonce;                    // Report.nonce
+  Extension extensions<4..2^16-1>; // Report.extensions
+  opaque verify_message<1..2^16-1>;
   EncryptedInputShare helper_share;
-  select (protocol) { // Protocol for the PPM task
-    case prio: PrioAggregateSubReq;
-    case hits: HitsAggregateSubReq;
-  }
-} AggregateSubReq;
+} VerifyStartSubReq;
+~~~
+
+Each sub-request includes the report's timestamp, nonce, and extensions (`time`,
+`nonce`, and `extensions` respectively) and a value called the verification
+message (`verify_message`), generated as described in the next paragraph. The
+first sub-request also carries the helper's encrypted input share
+(`helper_share`).
+
+The leader generates the verification message by decrypting its input share
+`leader_share`. To do so, it first looks up the HPKE secret key `sk` associated
+with `leader_share.aggregator_config_id`. Next, it runs
+
+~~~
+context = SetupBaseR(leader_share.enc, sk,
+                     "pda input share" || task_id || 0x00)
+input_share = context.Open(time || nonce || extensions,
+                           leader_share.payload)
 ~~~
 
-The `time`, `nonce`, and `extensions` fields have the same value as those in the
-report uploaded by the client. Similarly, the `helper_share` field is the
-`EncryptedInputShare` from the `Report` whose index in
-`Report.encrypted_input_shares` is equal to the index of `[aggregator]` in the
-task's aggregator endpoints. [OPEN ISSUE: We usually only need to send this in
-the first aggregate request. Shall we exclude it in subsequent requests
-somehow?] The remainder of the structure is dedicated to the protocol-specific
-request parameters.
+where `task_id`, `time`, `nonce`, and `extensions` are the fields of the report
+with the same name. If decryption succeeds, it runs the VDAF verify-start
+algorithm to compute its local VDAF state and the verification message:
 
-In order to provide replay protection, the leader is required to send aggregate
+~~~
+(state, VerifyStartSubReq.verify_message) = vdaf_start(
+      verify_param, aggregation_param, time || nonce, input_share)
```

Comment:
Do we not need a parameter indicating whether the entity running `vdaf_start` is the leader or helper, so that it can tell whether the input share is compressed or not (if the VDAF is Prio)? Or do we encode that information into `input_share`?

Reply:
It's encoded by the `input_share`. I know you didn't like this redundancy, but it makes the VDAF syntax a lot simpler.